File: cpp/src/lib/gherkin/rule_type.cpp
#include <unordered_map>

#include <gherkin/rule_type.hpp>

namespace gherkin {

std::string_view
to_string(rule_type r)
{
    static const std::unordered_map<rule_type, std::string_view> rmap = {
        { rule_type::none, "none" },
        { rule_type::e_o_f, "e_o_f" },
        { rule_type::empty, "empty" },
        { rule_type::comment, "comment" },
        { rule_type::tag_line, "tag_line" },
        { rule_type::feature_line, "feature_line" },
        { rule_type::rule_line, "rule_line" },
        { rule_type::background_line, "background_line" },
        { rule_type::scenario_line, "scenario_line" },
        { rule_type::examples_line, "examples_line" },
        { rule_type::step_line, "step_line" },
        { rule_type::doc_string_separator, "doc_string_separator" },
        { rule_type::table_row, "table_row" },
        { rule_type::language, "language" },
        { rule_type::other, "other" },
        { rule_type::gherkin_document, "gherkin_document" },
        { rule_type::feature, "feature" },
        { rule_type::feature_header, "feature_header" },
        { rule_type::rule, "rule" },
        { rule_type::rule_header, "rule_header" },
        { rule_type::background, "background" },
        { rule_type::scenario_definition, "scenario_definition" },
        { rule_type::scenario, "scenario" },
        { rule_type::examples_definition, "examples_definition" },
        { rule_type::examples, "examples" },
        { rule_type::examples_table, "examples_table" },
        { rule_type::step, "step" },
        { rule_type::step_arg, "step_arg" },
        { rule_type::data_table, "data_table" },
        { rule_type::doc_string, "doc_string" },
        { rule_type::tags, "tags" },
        { rule_type::description_helper, "description_helper" },
        { rule_type::description, "description" }
    };

    return rmap.at(r);
}

}

File: dart/lib/src/parser/rule_type.dart
/// The grammar rules recognized by the Gherkin parser.
///
/// The first group mirrors the lexical `TokenType`s one-to-one (the explicit
/// mapping lives in `TokenTypeExtension.toRuleType`); the remaining values are
/// the structural production rules of the Gherkin grammar.
enum RuleType {
  /// No rule; the initial/sentinel value.
  none,

  /// End-of-file token rule (`#EOF`).
  eof,

  /// Blank line token rule (`#Empty`).
  empty,

  /// Comment line token rule (`#Comment`).
  comment,

  /// Tag line token rule (`#TagLine`).
  tagLine,

  /// `Feature:` line token rule (`#FeatureLine`).
  featureLine,

  /// `Rule:` line token rule (`#RuleLine`).
  ruleLine,

  /// `Background:` line token rule (`#BackgroundLine`).
  backgroundLine,

  /// `Scenario:`/`Scenario Outline:` line token rule (`#ScenarioLine`).
  scenarioLine,

  /// `Examples:` line token rule (`#ExamplesLine`).
  examplesLine,

  /// Step line token rule (`#StepLine`).
  stepLine,

  /// Doc string delimiter token rule (`#DocStringSeparator`).
  docStringSeparator,

  /// Table row token rule (`#TableRow`).
  tableRow,

  /// `# language:` header token rule (`#Language`).
  language,

  /// Any other (description/free text) token rule (`#Other`).
  other,

  /// The whole document: `GherkinDocument := Feature?`.
  gherkinDocument,

  /// A feature:
  /// `Feature := FeatureHeader Background? ScenarioDefinition* Rule*`.
  feature,

  /// A feature header:
  /// `FeatureHeader := #Language? Tags? #FeatureLine DescriptionHelper`.
  featureHeader,

  /// A rule: `Rule := RuleHeader Background? ScenarioDefinition*`.
  rule,

  /// A rule header: `RuleHeader := #RuleLine DescriptionHelper`.
  ruleHeader,

  /// A background: `Background := #BackgroundLine DescriptionHelper Step*`.
  background,

  /// A (optionally tagged) scenario: `ScenarioDefinition := Tags? Scenario`.
  scenarioDefinition,

  /// A scenario:
  /// `Scenario := #ScenarioLine DescriptionHelper Step* ExamplesDefinition*`.
  scenario,

  /// A (optionally tagged) examples block:
  /// `ExamplesDefinition := Tags? Examples`.
  examplesDefinition,

  /// An examples block:
  /// `Examples := #ExamplesLine DescriptionHelper ExamplesTable?`.
  examples,

  /// An examples table: `ExamplesTable := #TableRow #TableRow*`.
  examplesTable,

  /// A step: `Step := #StepLine StepArg?`.
  step,

  /// A step argument: `StepArg := (DataTable | DocString)`.
  stepArg,

  /// A data table: `DataTable := #TableRow+`.
  dataTable,

  /// A doc string:
  /// `DocString := #DocStringSeparator #Other* #DocStringSeparator`.
  docString,

  /// One or more tag lines: `Tags := #TagLine+`.
  tags,

  /// A description helper:
  /// `DescriptionHelper := #Empty* Description?`.
  descriptionHelper,

  /// Free-text description: `Description := (#Other | #Comment)+`.
  description,
}
